pkg/authentication/context.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2024 Canonical Ltd.
+// SPDX-License-Identifier: AGPL-3.0
+
+package authentication
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/oauth2"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/monitoring"
+)
+
+type principalContextKey int
+
+const stateSize = 18
+
+var (
+	letterRunes         = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+	PrincipalContextKey principalContextKey
+)
+
+type Principal struct {
+	Subject string `json:"sub"`
+	Nonce   string `json:"nonce"`
+}
+
+func NewPrincipalFromClaims(c ReadableClaims) (*Principal, error) {
+	a := new(Principal)
+	if err := c.Claims(a); err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+func PrincipalContext(ctx context.Context, principal *Principal) context.Context {
+	parent := ctx
+	if ctx == nil {
+		parent = context.Background()
+	}
+
+	return context.WithValue(parent, PrincipalContextKey, principal)
+}
+
+func PrincipalFromContext(ctx context.Context) *Principal {
+	if ctx == nil {
+		return nil
+	}
+
+	return ctx.Value(PrincipalContextKey).(*Principal)
+}
+
+type OIDCProviderSupplier = func(ctx context.Context, issuer string) (*oidc.Provider, error)
+
+type OAuth2Context struct {
+	client   *oauth2.Config
+	verifier TokenVerifier
+
+	tracer  trace.Tracer
+	logger  logging.LoggerInterface
+	monitor monitoring.MonitorInterface
+	rand    *rand.Rand
+}
+
+func (o *OAuth2Context) LoginRedirect(w http.ResponseWriter, r *http.Request) {
+	_, span := o.tracer.Start(r.Context(), "authentication.OAuth2Context.LoginRedirect")
+	defer span.End()
+
+	nonce, _ := nonce(r)
+	state := o.state(stateSize)
+
+	SetNonceCookie(w, nonce)
+
+	http.Redirect(
+		w, r,
+		o.client.AuthCodeURL(state, oidc.Nonce(nonce), oauth2.SetAuthURLParam("audience", o.client.ClientID)),
+		http.StatusFound,
+	)
+}
+
+func (o *OAuth2Context) state(size int) string {
+	state := make([]rune, size)
+	for i := range state {
+		state[i] = letterRunes[rand.Intn(len(letterRunes))]
+	}
+	return string(state)
+}
+
+func nonce(r *http.Request) (string, error) {
+	marshal, err := json.Marshal(map[string]string{"referer": r.Referer(), "remote-address": r.RemoteAddr})
```

Comment:
Since we have a random generator for state, why don't we reuse that for nonce as well? Unless we have a use for these values.

Reply:
I would skip this additional change since it would require me to bring a bunch of changes to make it testable, and would cause an avalanche effect from addressing comments

Reply:
Same here. I have no problem with keeping this as it is for now, as long as we have a clear backlog of what needs to be done and that we eventually address this.

We should use random values for `nonce` and `state`. Using headers like `Referer` and `RemoteAddr` opens the window for attacks, because it is very easy for someone to tamper with them

Reply:
[Issue](https://github.com/canonical/identity-platform-admin-ui/issues/321) for this